### ISSUE TITLE
Camel - "Warning: Encountered two children with the same key" at routes node with multiple endpoints of the same kind

### ISSUE THREAD

**Issue by tadayosi** [hawtio/hawtio-next#368]
When I look at the Camel plugin and expand the `routes` folder, I can see there are two `to` endpoint nodes under `cron` in my example. Opening the browser console, I see the following error:

![image](https://github.com/hawtio/hawtio-next/assets/156692/67f59518-4fb3-40b9-9524-dec8355e243c)

> Warning: Encountered two children with the same key, `org.apache.camel-folder-CamelContexts-folder-SampleCamel-folder-routes-folder-cron-folder-to`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.

```
    at ul
    at TreeViewList (http://localhost:3000/hawtio/static/js/vendors-node_modules_patternfly_react-charts_dist_esm_index_js-node_modules_patternfly_react--46edeb.chunk.js:45861:7)
    at TreeView (http://localhost:3000/hawtio/static/js/vendors-node_modules_patternfly_react-charts_dist_esm_index_js-node_modules_patternfly_react--46edeb.chunk.js:45757:7)
    at li
    at TreeViewListItemBase (http://localhost:3000/hawtio/static/js/vendors-node_modules_patternfly_react-charts_dist_esm_index_js-node_modules_patternfly_react--46edeb.chunk.js:45901:5)
    at ul
    at TreeViewList (http://localhost:3000/hawtio/static/js/vendors-node_modules_patternfly_react-charts_dist_esm_index_js-node_modules_patternfly_react--46edeb.chunk.js:45861:7)
    at TreeView (http://localhost:3000/hawtio/static/js/vendors-node_modules_patternfly_react-charts_dist_esm_index_js-node_modules_patternfly_react--46edeb.chunk.js:45757:7)
    at li
    at TreeViewListItemBase (http://localhost:3000/hawtio/static/js/vendors-node_modules_patternfly_react-charts_dist_esm_index_js-node_modules_patternfly_react--46edeb.chunk.js:45901:5)
    at ul
    at TreeViewList (http://localhost:3000/hawtio/static/js/vendors-node_modules_patternfly_react-charts_dist_esm_index_js-node_modules_patternfly_react--46edeb.chunk.js:45861:7)
    at TreeView (http://localhost:3000/hawtio/static/js/vendors-node_modules_patternfly_react-charts_dist_esm_index_js-node_modules_patternfly_react--46edeb.chunk.js:45757:7)
    at li
    at TreeViewListItemBase (http://localhost:3000/hawtio/static/js/vendors-node_modules_patternfly_react-charts_dist_esm_index_js-node_modules_patternfly_react--46edeb.chunk.js:45901:5)
    at ul
    at TreeViewList (http://localhost:3000/hawtio/static/js/vendors-node_modules_patternfly_react-charts_dist_esm_index_js-node_modules_patternfly_react--46edeb.chunk.js:45861:7)
    at div
    at TreeViewRoot (http://localhost:3000/hawtio/static/js/vendors-node_modules_patternfly_react-charts_dist_esm_index_js-node_modules_patternfly_react--46edeb.chunk.js:46075:5)
    at TreeView (http://localhost:3000/hawtio/static/js/vendors-node_modules_patternfly_react-charts_dist_esm_index_js-node_modules_patternfly_react--46edeb.chunk.js:45757:7)
    at CamelTreeView (http://localhost:3000/hawtio/static/js/packages_hawtio_dist_index_js-data_image_png_base64_iVBORw0KGgoAAAANSUhEUgAAAAQAAAAECAYAAACp8-657d5d.chunk.js:104342:37)
    at div
    at div
    at SplitWrapper (http://localhost:3000/hawtio/static/js/vendors-node_modules_patternfly_react-charts_dist_esm_index_js-node_modules_patternfly_react--46edeb.chunk.js:289679:16)
    at Camel (http://localhost:3000/hawtio/static/js/packages_hawtio_dist_index_js-data_image_png_base64_iVBORw0KGgoAAAANSUhEUgAAAAQAAAAECAYAAACp8-657d5d.chunk.js:104413:7)
    at RenderedRoute (http://localhost:3000/hawtio/static/js/vendors-node_modules_react-router-dom_dist_index_js.chunk.js:5506:5)
    at Routes (http://localhost:3000/hawtio/static/js/vendors-node_modules_react-router-dom_dist_index_js.chunk.js:6076:5)
    at main
    at div
    at Page (http://localhost:3000/hawtio/static/js/vendors-node_modules_patternfly_react-charts_dist_esm_index_js-node_modules_patternfly_react--46edeb.chunk.js:34970:5)
    at HawtioPage (http://localhost:3000/hawtio/static/js/packages_hawtio_dist_index_js-data_image_png_base64_iVBORw0KGgoAAAANSUhEUgAAAAQAAAAECAYAAACp8-657d5d.chunk.js:106103:7)
    at RenderedRoute (http://localhost:3000/hawtio/static/js/vendors-node_modules_react-router-dom_dist_index_js.chunk.js:5506:5)
    at Routes (http://localhost:3000/hawtio/static/js/vendors-node_modules_react-router-dom_dist_index_js.chunk.js:6076:5)
    at Router (http://localhost:3000/hawtio/static/js/vendors-node_modules_react-router-dom_dist_index_js.chunk.js:6014:15)
    at BrowserRouter (http://localhost:3000/hawtio/static/js/vendors-node_modules_react-router-dom_dist_index_js.chunk.js:4190:5)
    at Hawtio (http://localhost:3000/hawtio/static/js/packages_hawtio_dist_index_js-data_image_png_base64_iVBORw0KGgoAAAANSUhEUgAAAAQAAAAECAYAAACp8-657d5d.chunk.js:106191:5)
```

**Comment by joshiraez** [comment]
This is tricky. I think there are some places in the code that I remember where we are under the assumption that names should be unique.

We could add an unique identifier to each node in the key prop to ensure no problems, but we would have to check the rest of the code to see everything is ok.

**Comment by tadayosi** [comment]
So, looking at the Source tab, we can see that the unique IDs are assigned to the steps from Camel already. We should use the same ID for our tree rendering.
```
<route xmlns="http://camel.apache.org/schema/spring" customId="true" id="cron">
        <from uri="quartz://cron?cron=0/10 * * * * ?"/>
        <setBody id="setBody1">
            <constant>Hello Camel! - cron</constant>
        </setBody>
        <to id="to1" uri="stream://out"/>
        <to id="to2" uri="mock://result"/>
    </route>
```

**Comment by joshiraez** [comment]
I believe this was the old behaviour that was causing @jsolovjo issues?

And for example the <from> doesn't have an id assigned. setBody shouldn't have the 1 as it's not needed. I'd have to see more examples but it's not as simple as just consuming them, when it's much easier to just use our own autonumerical on id coincidence.

I can try to go for a quick fix if this is desirable for 0.3.0 instead of bundling it together with https://github.com/hawtio/hawtio-next/issues/377 . It's just I feel using the generated camel id is actually a harder way than what we were thinking on doing before. Rather than that, I'd just go by the design we discussed where if there was a repeated id, we just added a sequential number to it.

On another note, this warning is just a warning set by react to tell you that your list might not be optimized. Given how we use the tree, there wont be any problems coming from having the same id other than slight performance issues at current I think. In other words, I believe it's not a blocker for 0.3.0 and would be good to bundle together with #377 to ensure there are no more reimplementations of the tree and have an unified design for it.

**Comment by lhein** [comment]
Camel sets ID values automatically when firing up Camel Routes. There will always be an ID value assigned to every element in a route because of that. The way it generates the ID is to take the element name and adding a number which is increasing with every found element with the same name. We need to use the ID values rather than inventing our own.

**Comment by joshiraez** [comment]
A few counterpoints to that:

1) You can see from the example that not all elements in the route have ids
2) We are subject to that Id definition to change any time without us having no control over that. If https://github.com/hawtio/hawtio-next/issues/387 is anything to go by, or other Camel related projects, we shouldn't leisurely rely on Camel specifics as we are quite much subject to them doing breaking changes that might hinder our application. Ids are only an internal used by the HTML rendering and tests, so I see no need to have to use the ids provided by camel (as we are not doing it anyway).
3) This doesn't resolve the issue that we might have duplicated ids in the JMX, which could happen as well (https://github.com/hawtio/hawtio-next/issues/377 which is the issue that is being solved together with this one).
4) This caused issues for testing in the past if I'm not mistaken. That's why we went with the current ids.

We need to remember that our application is not only for camel. I understand your point Lars but I feel like just following it blindly without more info will cause more problems than solutions. Plus it's actually harder to retrieve those element ids than to just do our own (which is something we already have to do), and our ids are not used like in camel, just being HTML helpers for React and testing, so the use cases are totally different.

**Comment by phantomjinx** [comment]
Actually I think during the course of sorting the https://github.com/hawtio/hawtio-next/issues/387 issue is fixing this error too

**Comment by joshiraez** [comment]
How is not #377 related to #368 if #377 was discovered while testing for ids, which is a behaviour of #368, and the part that I rollbacked from https://github.com/hawtio/hawtio-next/pull/380 was the part that fixed the wrong ids, and we decided that we needed a more in-depth look instead of having the ids be generated around when appending parents and childrens, and taking into account duplicates as to they have to be immutable and change the ids in case there are repeated nodes (which is the cause of #377, and was one of the things fixed with the `-folder` id that was done in https://github.com/hawtio/hawtio-next/pull/313 and later rediscovered)

We have to take into account that even right now, things like this test are broken: https://github.com/hawtio/hawtio-next/pull/380/commits/b8c86143270f9d18761464fc75b526a8e8d0b8fc as it didn't start to fail until I did my first batch of changes that I had to rollback.

For example, it causes issues like https://github.com/hawtio/hawtio-next/pull/392/files#diff-694a31e072413ae06e0ae5f11c3ea3e0ceabbfad98dab29d3fa992b2be4c99fcR96 as that test shouldn't be passing in the first instance (those MBean nodes are actually childrens of a context, so we know that with the way that we generate ids now that couldn't ever pass)

All this was discussed in Friday and Monday already. But I have no problems putting together all the context so we can have a better discussion on what we want.

That not to say that I struggle to see what's the advantage of having our own IDs that we use for our own things, instead of going with Camel's, when I think it was a source of a bug in the past.

But my point is... I feel like there is a sense of urgency for something that was already discussed were related, and it has no Milestone associated. If we want it for 0.3.0 I can put back the quick fix I did originally and finish this issue, then keep working on #377. But I'd like to know why the hurry and change on priorities over things we have already discussed?
